src/diff_walker.rs

```diff
@@ -27,21 +27,43 @@ impl DiffWalker {
         if let (Some(lhs_any_of), Some(rhs_any_of)) =
             (&mut lhs.subschemas().any_of, &mut rhs.subschemas().any_of)
         {
-            lhs_any_of.sort_by_cached_key(|x| format!("{x:?}"));
-            rhs_any_of.sort_by_cached_key(|x| format!("{x:?}"));
+            match (lhs_any_of.len(), rhs_any_of.len()) {
+                (l, r) if l <= r => {
+                    lhs_any_of.append(&mut vec![Schema::Bool(false); r - l]);
+                }
+                (l, r) => {
+                    rhs_any_of.append(&mut vec![Schema::Bool(false); l - r]);
+                }
+            }
 
-            for (i, (lhs_inner, rhs_inner)) in
-                lhs_any_of.iter_mut().zip(rhs_any_of.iter_mut()).enumerate()
-            {
+            let len = lhs_any_of.len();
+            let mut mat = pathfinding::matrix::Matrix::new(len, len, 0i32);
+            for (i, l) in lhs_any_of.iter_mut().enumerate() {
+                for (j, r) in rhs_any_of.iter_mut().enumerate() {
+                    let mut walker = DiffWalker {
+                        changes: vec![],
+                        lhs_root: self.lhs_root.clone(),
+                        rhs_root: self.rhs_root.clone(),
+                    };
+                    walker.diff(
+                        "",
+                        &mut l.clone().into_object(),
+                        &mut r.clone().into_object(),
+                    )?;
+                    mat[(i, j)] = i32::try_from(walker.changes.len()).expect("too many changes");
+                }
+            }
```

Comment:
It would be more convenient for `Diffwalker::diff` to return `Vec<Change>` instead of mutating internal changes. By doing so we can reuse `Diffwalker` and avoid cloning `RootSchema` here

Reply:
there is generally way too much cloning going on. I haven't wrapped my head around it yet, but I already have a problem with the fact that we cannot just pass an immutable reference when we recurse into schemas

Reply:
I would not change it for now, I want to figure out how we can eliminate as many clones as possible. changing DiffWalker now might make it harder to remove clone() in a hotter codepath.

Reply:
@6293 actually, I think we should change the design of diff walker in the following way:

1. instead of a vector of changes, it is generic over the collection that holds the changes. this can be implemented by making `DiffWalker` take a callback for appending a change
2. at this point I think `DiffWalker` should have a proper `new` constructor as well, because we instantiate it in two places
3. in the public API, we can pass `myvector.push` as callback. but here, we can increment an integer and discard the change
4. if there are too many changes, we clip them at `i32::MAX`